platforms/allwinner-d1/src/dmac/mod.rs

```diff
@@ -135,6 +242,39 @@ impl Channel {
         }
     }
 
+    pub async unsafe fn run_descriptor(&mut self, desc: NonNull<Descriptor>) {
+        // mark the channel as in-flight.
+        let prev_state = self
+            .channel
+            .state
+            .fetch_or(ChannelState::IN_FLIGHT, Ordering::AcqRel);
+        assert_eq!(
+            prev_state & ChannelState::IN_FLIGHT,
+            0,
+            "cannot start DMA transfer on a channel that already has an in-flight transfer",
+        );
+
+        // pre-subscribe to the waitcell to ensure our waker is registered
+        // before starting the DMA transfer.
+        let wait = self.channel.waker.subscribe().await;
+
+        // actually start the DMA transfer.
+        self.start_descriptor(desc);
+
+        // wait for the DMA transfer to complete.
```

Comment:
One thing we still haven't tackled here is that dropping/cancelling/forgetting a DMA future will lead to Very Bad Bonus Memory.

for dropping/cancelling, we could either force-stop the dma transfer (with a manual future that impls Drop), but solving this for `forget` means that the transfer future needs to take owership of the destination buffer, in such a way that forgetting the future also permanently leaks the destination buffer (which might be a whole ass peripheral?).

I'm not against saying "using `forget` in kernel futures is not allowed".

You don't need to fix that issue now, just reminding it exists.

Reply:
yeah, i was also thinking about this. i would like to figure out a safe API for it too, but this is why i left the `Channel::run_descriptor` method unsafe for now. we should probably add some docs...

Reply:
Ah, that's a good point, I think if you mention it in the function's SAFETY docs then we're all good (or at least better off than before!), tho the parent (which might be a safe future) can still be cancelled.

Reply:
yeah, we should have some safety docs for this, for sure. i can add them now!

Reply:
left a bunch of comments!

Reply:
we can probably stick something in the `run_descriptor` future that cancels the DMA when it's dropped, making this cancel-safe. but, there's basically no way to make this safe in the face of `mem::forget`, so i think we can just keep the function `unsafe` and document "uwu no forgetty". 

honestly i don't think making driver-facing internal kernel APIs safe with regards to `mem::forget`ting futures is one of my biggest priorities, since it's extremely hard, and if anyone is writing code that's randomly forgetting futures without dropping them, something has gone Really Fucking Wrong somewhere. i would _like_ to be resilient in this case, but i don't think it's always even possible without extreme contortion...

Reply:
at the level of the driver that's running DMA xfers i would expect buffers to be owned either by the driver or some external thing like a shared BBQueue, but inside of the driver i think it's fair to say "driver tasks that `mem::forget` DMA futures deserve whatever happens to them". which, i suppose, means we can't really make these APIs safe. which i'm willing to accept.